preflight-packages/examples.jetstack.io/pods_basic/pods.rego

```diff
@@ -3,67 +3,74 @@ package pods
+	container := pod.spec.containers[_]
+	# test if the limits are not set
+	not container.resources.limits.cpu
+	# bind a message for reporting
+	message := sprintf("container '%s' in pod '%s' in namespace '%s' is missing a cpu limit", [container.name, pod.metadata.name, pod.metadata.namespace])
 }
-init_containers_without_cpu_limits[container_name] {
-        pod := pods.items[_]
-        container := pod.spec.initContainers[_]
-        container_name = format_container(pod, container)
-        cpu_limit_unset(container)
+preflight_container_cpu_limit[message] {
+	# find all initContainers in all pods
+	pod := pods.items[_]
+	container := pod.spec.initContainers[_]
+	# test if the limits are not set
+	not container.resources.limits.cpu
+	# bind a message for reporting
+	message := sprintf("init container '%s' in pod '%s' in namespace '%s' is missing a cpu limit", [container.name, pod.metadata.name, pod.metadata.namespace])
 }
 
 # Rule 'container_mem_limit'
-default preflight_container_mem_limit = false
-preflight_container_mem_limit {
-        count(containers_without_memory_limits) == 0
-        count(init_containers_without_memory_limits) == 0
-}
-memory_limit_unset(container) {
-        not container.resources.limits.memory
+preflight_container_mem_limit[message] {
+	# find all containers in all pods
+	pod := pods.items[_]
+	container := pod.spec.containers[_]
+	# test if the limits are not set
+	not container.resources.limits.memory
+	# bind a message for reporting
+	message := sprintf("container '%s' in pod '%s' in namespace '%s' is missing a memory limit", [container.name, pod.metadata.name, pod.metadata.namespace])
 }
-containers_without_memory_limits[container_name] {
-        pod := pods.items[_]
-        container := pod.spec.containers[_]
-        container_name = format_container(pod, container)
-        memory_limit_unset(container)
-}
-init_containers_without_memory_limits[container_name] {
-        pod := pods.items[_]
-        container := pod.spec.initContainers[_]
-        container_name = format_container(pod, container)
-        memory_limit_unset(container)
+preflight_container_mem_limit[message] {
+	# find all initContainers in all pods
+	pod := pods.items[_]
+	container := pod.spec.initContainers[_]
+	# test if the limits are not set
+	not container.resources.limits.memory
+	# bind a message for reporting
+	message := sprintf("init container '%s' in pod '%s' in namespace '%s' is missing a memory limit", [container.name, pod.metadata.name, pod.metadata.namespace])
 }
 
 # Rule 'explicit_image_tag'
-default preflight_explicit_image_tag = false
-preflight_explicit_image_tag {
-        count(containers_without_explicit_tag) == 0
+preflight_explicit_image_tag[message] {
+	# find all containers in all pods
+	pod := pods.items[_]
+	container := pod.spec.containers[_]
+	# validate that the image value contains an explicit tag
+	{ re_match(`latest$`, container.image),
+	re_match(`^[^:]+$`, container.image) } & { true } != set()
```

Comment:
I'm slightly confused by this, why is the ` & { true }` required?

Reply:
This is just a compact way to write OR in rego... :grimacing: 

https://stackoverflow.com/a/58325979/1510063 outlines some other options

Reply:
& is intersection, so { true, x } & { true } == { true }

Which is not empty, and so the message is bound as that's a violation as one of the image requirements was not met

Reply:
We could use incremental rules if you'd rather.

Reply:
> & is intersection

Now it makes sense, I was being slow and thinking of it as a binary thing.

Reply:
If we have more queries about it I might move to more verbose incremental rules - they have the added benefit of more descriptive error messages too.